### PR TITLE
Add a user guide for the cluster registry.

### PR DESCRIPTION
**Pull request by perotinus**
/sig multicluster

Addresses #89.

### CODE REVIEW

File: docs/userguide.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+# User Guide
+
+## What is the cluster registry?
+
+The cluster registry is a Kubernetes-style API and API server that provides an
+endpoint for interacting with a list of clusters and associated metadata. If it
+helps, you can think of the cluster registry as a hosted kubeconfig file.
+However, since it's a Kubernetes API server, the cluster registry allows custom
```
Comment (madhusudancs) [inline docs/userguide.md]: Aren't all these the properties of a kubernetes-style API, not the API server?

File: docs/userguide.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+Kubernetes-style generated client libraries, and supports having controllers
+watch for updates.
+
+## Deploying a cluster registry
+
+The `crinit` tool allows you to deploy a cluster registry into an existing
+Kubernetes cluster. Run the following commands to download the latest nightly
+build of the tool:
+
+> Currently the tool is only released for Linux 64-bit. You'll need to built it
```
Comment (madhusudancs) [inline docs/userguide.md]: s/built/build/

File: docs/userguide.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+curl -O http://storage.googleapis.com/crreleases/nightly/$LATEST/clusterregistry-$PACKAGE.tar.gz
+tar xzf clusterregistry-$PACKAGE.tar.gz
+```
+
+### Standalone
+
+You can deploy the cluster registry as a standalone API server like so:
+
+```sh
+./crinit standalone <cluster_registry_context> --host-cluster-context=<your_cluster_context>
```
Comment (madhusudancs) [inline docs/userguide.md]: You probably need to `cd` into the extracted directory at some point? Without `cd` the flow kind of breaks for people following along?

File: docs/userguide.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+```
+
+where `cluster_registry_context` is the name you want to give the new context in
+your kubecontext file that will point to your cluster registry and
```
Comment (madhusudancs) [inline docs/userguide.md]: kubeconfig file.

File: docs/userguide.md (hunk 3 above)
Comment (madhusudancs) [reply docs/userguide.md]: It's best to call the positional argument the cluster registry instance name. It has implications far beyond a context name in the local kubeconfig file. For example, it is used for the deployment and service names so the name should conform to those rules etc.

File: docs/userguide.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+`host-cluster-context` is a context entry in your kubeconfig file. When the
```
Comment (madhusudancs) [inline docs/userguide.md]: `host-cluster-context` deserves some more explanation.

File: docs/userguide.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+> will fail because the cluster registry will not be able to verify the user.
+> This is currently being addressed.
+
+You can deploy the cluster registry as an aggregated API server like so:
+
+```sh
+./crinit aggregated <cluster_registry_context> --host-cluster-context=<your_cluster_context>
+```
+
+where `host-cluster-context` is a context entry in your kubeconfig file. When
```
Comment (madhusudancs) [inline docs/userguide.md]: This deserves more explanation too.

File: docs/userguide.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+In these examples, `context` is either the `cluster_registry_context` if you did
+a standalone deployment, or the `host_cluster_context` if you did an aggregated
+deployment.
+
+Try creating a cluster:
+
+```sh
+kubectl apply -f - --context <context> <<EOF
+kind: Cluster
+apiVersion: clusterregistry.k8s.io/v1alpha1
```
Comment (madhusudancs) [inline docs/userguide.md]: Are we approved for this name? AFAIU, `X.k8s.io` is only for k8s APIs, not for the k8s-style APIs. Are @kubernetes/api-approvers Ok with this?

File: docs/userguide.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+certificates and client authentication/authorization yourself. You can use any
+authentication mode supported by the `clusterregistry` command, but you will
+probably need to use an
+[authorizing webhook](https://kubernetes.io/docs/admin/authorization/webhook/),
+since the cluster registry does not provide RBAC support inherently.
+
+#### etcd
+
+`crinit` by default deploys `etcd` in a container in the same pod as the
+`clusterregistry` binary. This makes it easy to get started, but is not
```
Comment (madhusudancs) [inline docs/userguide.md]: s/binary/container/

File: docs/userguide.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+necessarily a good strategy for deploying a cluster registry for production use.
+You may want to look into the
+[etcd operator](https://github.com/coreos/etcd-operator) for production
+deployments.
+
+## Interacting with the cluster registry
+
+### kubectl
+
+The cluster registry is a Kubernetes API server, and you can interact with it
```
Comment (madhusudancs) [inline docs/userguide.md]: Kubernetes-style API server

File: docs/userguide.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+using standard `kubectl` commands. It provides one API type, `clusters`, which
+you can create, get, list and delete like any other Kubernetes Object. See
```
Comment (madhusudancs) [inline docs/userguide.md]: s/Object/object/

File: docs/userguide.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+### Generated Go client
+
+There is a generated Go client library for the cluster registry in
+[/pkg/client](/pkg/client). You can vendor in the cluster registry repository
+and use the client library directly from your Go code.
+
+### OpenAPI spec
+
+There is an OpenAPI spec file provided [here](/api/swagger.json). You can use
```
Comment (madhusudancs) [inline docs/userguide.md]: Is the URL legit? If it isn't I would just drop here and tell users how to get the spec from their API server.

File: docs/userguide.md (hunk 2 above)
Comment (perotinus) [reply docs/userguide.md]: Done.

File: docs/userguide.md (hunk 1 above)
Comment (perotinus) [reply docs/userguide.md]: That's a good point. Changed.

File: docs/userguide.md (hunk 3 above)
Comment (perotinus) [reply docs/userguide.md]: There is no extracted directory: the command is extracted into the current working directory.

The point about the positional argument makes a lot of sense. Fixed.

File: docs/userguide.md (hunk 4 above)
Comment (perotinus) [reply docs/userguide.md]: Done.

File: docs/userguide.md (hunk 5 above)
Comment (perotinus) [reply docs/userguide.md]: Done.

File: docs/userguide.md (hunk 6 above)
Comment (perotinus) [reply docs/userguide.md]: Done.

File: docs/userguide.md (hunk 7 above)
Comment (perotinus) [reply docs/userguide.md]: No, we haven't gone through that process yet. That said, I believe that the Cluster API group is also using a `k8s.io` name, as is ServiceCatalog; and this is still pre-alpha, so we can change this later without breaking any promises about support.

Filed #121 to track.

File: docs/userguide.md (hunk 8 above)
Comment (perotinus) [reply docs/userguide.md]: Done.

File: docs/userguide.md (hunk 9 above)
Comment (perotinus) [reply docs/userguide.md]: Done

File: docs/userguide.md (hunk 11 above)
Comment (perotinus) [reply docs/userguide.md]: It is, surprisingly: GitHub takes absolute path links without schemes and references them from the root of the repository.

File: docs/userguide.md (hunk 10 above)
Comment (perotinus) [reply docs/userguide.md]: Done.

**Comment by perotinus** [comment]
Thanks, @madhusudancs!

File: docs/userguide.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+### Standalone
+
+You can deploy the cluster registry as a standalone API server like so:
+
+```sh
+./crinit standalone <cluster_registry_instance_name> --host-cluster-context=<your_cluster_context>
+```
+
+where `cluster_registry_instance_name` is the name you want to give this cluster
+registry instance the and `host-cluster-context` is a context entry in your
```
Comment (madhusudancs) [inline docs/userguide.md]: s/the and/and/

File: docs/userguide.md (hunk 3 above)
Comment (madhusudancs) [reply docs/userguide.md]: > There is no extracted directory: the command is extracted into the current working directory.

I see Ok. Thanks for the explanation.

File: docs/userguide.md (hunk 7 above)
Comment (madhusudancs) [reply docs/userguide.md]: Fair enough. Thanks for opening an issue to track this.

File: docs/userguide.md (hunk 11 above)
Comment (madhusudancs) [reply docs/userguide.md]: Oh wow! Didn't know this. TIL.

File: docs/userguide.md (hunk 12 above)
Comment (perotinus) [reply docs/userguide.md]: Thanks, fixed.

**Pull request merged**